ntp-daemon/src/peer_manager.rs

```diff
@@ -78,41 +153,45 @@ impl<C: NtpClock> Peers<C> {
         }
     }
 
-    async fn add_peer_internal(&mut self, config: Arc<PeerConfig>) -> JoinHandle<()> {
+    /// Add a single peer to an existing pool
+    async fn add_pool_peer_internal(
+        &mut self,
+        address: NormalizedAddress,
+        mut cached: CachedPoolAddresses,
+    ) -> JoinHandle<()> {
         let index = self.indexer.get();
-        let addr = loop {
-            let host = match &*config {
-                PeerConfig::Standard(StandardPeerConfig { addr, .. }) => {
-                    debug!(unresolved = ?&addr, "lookup host");
-                    addr.lookup_host().await.map(|mut i| i.next())
-                }
-
-                PeerConfig::Pool(PoolPeerConfig { addr, .. }) => {
-                    debug!(unresolved = ?&addr, "lookup host");
-                    addr.lookup_host().await.map(|mut i| i.next())
-                }
-            };
 
-            match host {
-                Ok(Some(addr)) => {
-                    debug!(resolved=?addr, "resolved peer");
-                    break addr;
-                }
-                Ok(None) => {
-                    warn!("Could not resolve peer address, retrying");
-                    tokio::time::sleep(NETWORK_WAIT_PERIOD).await
-                }
-                Err(e) => {
-                    warn!(error = ?e, "error while resolving peer address, retrying");
-                    tokio::time::sleep(NETWORK_WAIT_PERIOD).await
+        // socket addresses of the peers of this pool that are currently active
+        let active_pool_peers = self.peers.values().filter_map(|p| match &p.peer_address {
+            PeerAddress::Peer { .. } => None,
+            PeerAddress::Pool {
+                address: peer_address,
+                socket_address,
+                ..
+            } => {
+                if &address == peer_address {
+                    Some(*socket_address)
+                } else {
+                    None
                 }
             }
+        });
+
+        // if no socket addresses are cached, do a new DNS resolve
+        let addr = match cached.find_additional(&address, active_pool_peers).await {
+            Some(addr) => addr,
+            None => todo!("the pool is now smaller than max_peers; what to do?"),
```

Comment:
I think we're kind of stuck at this point. We want to add another peer, but there just aren't any left in the pool. Any thoughts?

Reply:
I see two potential options. My preference is to just print an error here letting the user know they borked up their configuration and letting them fix it through a restart.

We could instead consider starting a background process that tries to fix this after some time has passed (together ofcourse with a warning or error), but I think that is a bad idea as it is fairly complex and adds a significant risk of actually starting to ddos the pools dns servers should this situation ever occur.

Reply:
sounds good! I'll add some error logging and then this function needs to return an `Option`